p2p/cryptoBox/box.go

```diff
@@ -0,0 +1,159 @@
+package cryptoBox
+
+import (
+	crypto_rand "crypto/rand"
+	"errors"
+	"fmt"
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"golang.org/x/crypto/nacl/box"
+	"io"
+)
+
+const (
+	keySize = 32 // non-configurable, expected by NaCl
+	nonceSize = 24 // non-configurable, expected by NaCl
+)
+
+type Key interface {
+	Bytes() []byte
+	Raw() *[keySize]byte
+	String() string
+	Pretty() string
+}
+
+type PrivateKey interface {
+	Key
+}
+
+type PublicKey interface {
+	Key
+}
+
+type SharedSecret interface {
+	Key
+	Seal(message []byte) (out []byte)
+	Open(encryptedMessage []byte) (out []byte, err error)
+}
+
+type key struct {
+	bytes *[keySize]byte
```

Comment:
this might cause unnecessary boxing / unboxing, I'd rather you save a fixed size array and when needed send its address using &

Reply:
you cant really know where it is allocated without some research, and keep in mind these things can change between go versions.
i would be careful  investing time on these sort of things without a concrete problem to solve

Reply:
I agree with @almogdepaz that this is on the verge of premature optimization.

However... Thanks to your comment, @antonlerner, I deepened my understanding of memory allocation in Go. So first of all - thanks. I've read [this](https://segment.com/blog/allocation-efficiency-in-high-performance-go-services/) on the topic and as a result I did change some code (including what you commented on) to use less pointers.

Although we now use less pointers, much of these variables escape to the heap anyway:
```
$ go build -gcflags '-m' p2p/cryptoBox/box.go 2>&1 >/dev/null | grep "escapes to heap"
p2p/cryptoBox/box.go:47:9: &k.bytes escapes to heap
p2p/cryptoBox/box.go:51:16: k.bytes escapes to heap
p2p/cryptoBox/box.go:60:45: nonce escapes to heap
p2p/cryptoBox/box.go:68:42: nonce escapes to heap
p2p/cryptoBox/box.go:76:25: error(&errors.errorString literal) escapes to heap
p2p/cryptoBox/box.go:76:25: &errors.errorString literal escapes to heap
p2p/cryptoBox/box.go:88:12: key literal escapes to heap
p2p/cryptoBox/box.go:88:27: key literal escapes to heap
p2p/cryptoBox/box.go:94:2: sharedSecret escapes to heap
p2p/cryptoBox/box.go:120:29: l escapes to heap
p2p/cryptoBox/box.go:120:29: keySize escapes to heap
p2p/cryptoBox/box.go:103:31: keySize escapes to heap
p2p/cryptoBox/box.go:103:31: mSize escapes to heap
p2p/cryptoBox/box.go:134:28: error(&errors.errorString literal) escapes to heap
p2p/cryptoBox/box.go:134:28: &errors.errorString literal escapes to heap
```

Doing any further optimization would cause code to be less readable and I won't do it without knowing that it's justified after running some analysis tools when we have a working MVP.